python/shotgun_data/shotgun_data_retriever.py

```diff
@@ -0,0 +1,723 @@
+                            with the image data loaded.
+
+        :returns:           A unique task id representing this request.
+        """
+        if not self._task_manager:
+            return
+
+        # always add check for thumbnail already downloaded:
+        check_task_id = self._task_manager.add_task(self._task_check_thumbnail,
+                                                    priority = ShotgunDataRetriever._CHECK_THUMB_PRIORITY,
+                                                    group = self._bg_tasks_group,
+                                                    url = url,
+                                                    load_image = load_image)
+
+        # add download thumbnail task
+        dl_task_id = self._task_manager.add_task(self._task_download_thumbnail,
+                                                 upstream_task_ids = [check_task_id],
+                                                 priority = ShotgunDataRetriever._DOWNLOAD_THUMB_PRIORITY,
+                                                 group = self._bg_tasks_group,
+                                                 url = url,
+                                                 entity_type = entity_type, 
+                                                 entity_id = entity_id,
+                                                 field = field,
+                                                 load_image = load_image)
+
+        # all results for requesting a thumbnail should be returned with the same id so use
+        # a mapping to track the 'primary' task id:
+        self._thumb_task_id_map[dl_task_id] = check_task_id
+
+        return str(check_task_id)
+
+
+
+    # ------------------------------------------------------------------------------------------------
+    # Background task management and methods
+
+    def _task_get_schema(self, project_id):
+        """
+        Method that gets executed in a background task/thread to retrieve the fields
+        and types schema from Shotgun
+
+        :param project_id:  The id of the project to query the schema for or None to 
+                            retrieve for all projects
+        :returns:           Dictionary containing the 'action' together with the schema 
+                            fields and types
+        """
+        if project_id is not None:
+            project = {"type": "Project", "id": project_id}
+        else:
+            project = None
+
+        # read in details about all fields
+        sg_field_schema = self._shotgun_connection.schema_read(project)
+
+        # and read in details about all entity types
+        sg_type_schema = self._shotgun_connection.schema_entity_read(project)
+
+        # need to wrap it in a dict not to confuse pyqt's signals and type system
+        return {"action":"schema", "fields":sg_field_schema, "types":sg_type_schema}
+
+    def _task_execute_find(self, *args, **kwargs):
```

Comment:
I think clarity wins here - having separate methods for each shotgun call makes the code easier to read and debug, especially since the callstack you'd get from one of these methods is going to be minimal (as it's in a thread) - having an exception from a _task_execute_crud is going to be a lot harder to figure out the cause than if it says _task_execute_update for example.